src/main/java/frc/robot/subsystems/Elevator.java

```diff
@@ -0,0 +1,107 @@
+/*----------------------------------------------------------------------------*/
+/* Copyright (c) 2018-2019 FIRST. All Rights Reserved.                        */
+/* Open Source Software - may be modified and shared by FRC teams. The code   */
+/* must be accompanied by the FIRST BSD license file in the root directory of */
+/* the project.                                                               */
+/*----------------------------------------------------------------------------*/
+
+package frc.robot.subsystems;
+import com.ctre.phoenix.motorcontrol.can.WPI_TalonFX;
+
+import edu.wpi.first.math.MathUtil;
+import edu.wpi.first.math.controller.PIDController;
+import edu.wpi.first.wpilibj.DigitalInput;
+import edu.wpi.first.wpilibj.DutyCycleEncoder;
+import edu.wpi.first.wpilibj2.command.SubsystemBase;
+import frc.robot.constants.Constants;
+import frc.robot.constants.ElevatorConstants;
+import frc.robot.controls.Operator;
+import frc.robot.util.MotorFactory;
+
+public class Elevator extends SubsystemBase {
+
+  private final WPI_TalonFX m_elevatorMotor; //do I define these are private final? 
+  private final DigitalInput m_bottomLimitSwitch;
+  private final DigitalInput m_topLimitSwitch; 
+  private final PIDController m_elevatorPID; 
+  private final DutyCycleEncoder m_absoluteSpoolEncoder; 
+
+  public Elevator() {
+    this(
+      MotorFactory.createTalonFX(ElevatorConstants.kElevatorMotor, Constants.kRioCAN),
+      new DigitalInput(-1),
+      new DigitalInput(-1),
+      new DutyCycleEncoder(-1)
```

Comment:
Last year we made the subsystems this way because we had to mock them. I'm not sure that's necessary this year. Also, the limit switch ports and encoder ports should be constants.

Reply:
Code Change incorporated

Reply:
Absolutely not necessary.

Just have one constructor. It should set the member values using defined port constants.

    m_elevatorMotor = MotorFactory.createTalonFX(ElevatorConstants.kElevatorMotor, Constants.kRioCAN); 
    m_bottomLimitSwitch = new DigitalInput(kBottomLimitSwitchPort); 
    m_topLimitSwitch = new DigitalInput(kTopLimitSwitchPort);
    m_absoluteSpoolEncoder = new DutyCycleEncoder(kSpoolEncoder); 

Those assignment could happen where the member variables are declared.

The DutyCycleEncoder allows you to set an offset and a scale.
See DutyCycleEncoder.setPositionOffset() and .setDistancePerRotation().
They should be set in the constructor.

You should add a close method to close the digital ports. It can be called during testing.

    public void close() {
        // close the ports
        m_bottomLimitSwitch.close();
        m_topLimitSwitch.close();
        m_absoluteSpoolEncoder.close();
    }

Reply:
Change added.